api/v1alpha1/warehouse_types.go

```diff
@@ -267,6 +284,127 @@ type WarehouseStatus struct {
+	// +optional
+	Commits []DiscoveredCommit `json:"commits,omitempy" protobuf:"bytes,2,rep,name=commits"`
+}
+
+// DiscoveredCommit represents a commit discovered by a Warehouse for a
+// GitSubscription.
+type DiscoveredCommit struct {
+	// ID is the identifier of the commit. This typically is a SHA-1 hash.
+	// +kubebuilder:validation:MinLength=1
+	ID string `json:"id,omitempty" protobuf:"bytes,1,opt,name=id"`
+
+	// Branch is the branch in which the commit was found. This field is
+	// optional, and populated based on the CommitSelectionStrategy of the
+	// GitSubscription.
+	// +optional
+	Branch string `json:"branch,omitempty" protobuf:"bytes,2,opt,name=branch"`
+
+	// Tag is the tag that resolved to this commit. This field is optional, and
+	// populated based on the CommitSelectionStrategy of the GitSubscription.
+	// +optional
+	Tag string `json:"tag,omitempty" protobuf:"bytes,3,opt,name=tag"`
+
+	// Subject is the subject of the commit (i.e. the first line of the commit
+	// message).
+	// +optional
+	Subject string `json:"subject,omitempty" protobuf:"bytes,4,opt,name=subject"`
+
+	// Author is the author of the commit.
+	// +optional
+	Author string `json:"author,omitempty" protobuf:"bytes,5,opt,name=author"`
+
+	// CreatedAt is the time at which the commit or annotated tag was created.
+	// +optional
+	CreatedAt *metav1.Time `json:"createdAt,omitempty" protobuf:"bytes,6,opt,name=createdAt"`
+}
+
+// ImageDiscoveryResult represents the result of an image discovery operation
+// for an ImageSubscription.
+type ImageDiscoveryResult struct {
+	// RepoURL is the repository URL of the image, as specified in the
+	// ImageSubscription.
+	// +kubebuilder:validation:MinLength=1
+	RepoURL string `json:"repoURL" protobuf:"bytes,1,opt,name=repoURL"`
+
+	// Images is a list of images discovered by the Warehouse for the
+	// ImageSubscription. An empty list indicates that the discovery operation
+	// was successful, but no images matching the ImageSubscription criteria were
+	// found.
+	// +optional
+	Images []DiscoveredImage `json:"images" protobuf:"bytes,2,rep,name=images"`
+}
+
+// DiscoveredImage represents an image discovered by a Warehouse for an
+// ImageSubscription.
+type DiscoveredImage struct {
+	// Tag is the tag of the image.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=128
+	// +kubebuilder:validation:Pattern=`^[\w.\-\_]+$`
+	Tag string `json:"tag" protobuf:"bytes,1,opt,name=tag"`
+	// Digest is the digest of the image.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=`^[a-z0-9]+:[a-f0-9]+$`
+	Digest string `json:"digest" protobuf:"bytes,2,opt,name=digest"`
+	// Platform is the target platform of the image. This field is optional, and
+	// only populated if the ImageSubscription specifies a Platform.
+	// +optional
+	Platform string `json:"platform,omitempty" protobuf:"bytes,3,opt,name=platform"`
```

Comment:
I wonder slightly if this is important information and I also wonder if it could lead to confusion.

In terms of image selection, when a tag points to a multi-arch image's manifest list (v2) or index (OCI) it is the digest of the list/index that is returned (as long as that list or index does indeed reference an image having the desired architecture). It's _not_ the digest of the single-arch image that is returned. This behavior makes sense because 1. That's not what the tag points to and 2. This behavior leaves containerd free to self-select the correct platform out of the multiple ones available per usual.

So the confusion that could arise here is that the digest may reference a multi-arch image while the platform field, with a value copied over from the platform constraint, may promote the idea that this image is for a specific platform only.

Reply:
In hindsight, I think I agree. The reason I added it was because I wanted to allow the UI to potentially information about the architecture constraint. However, this is likely better to add one level higher as it will be the same for every listed result anyway.

Reply:
Ah... sure... I agree moving it up a level would clear up the confusion.